drivers/modem/ublox-sara-r4.c

```diff
@@ -0,0 +1,1854 @@
+		goto error;
+	}
+
+#if defined(CONFIG_BOARD_PARTICLE_BORON)
+	/* use external SIM */
+	ret = send_at_cmd(NULL, "AT+UGPIOC=23,0,0", MDM_CMD_TIMEOUT);
+	if (ret < 0) {
+		LOG_ERR("AT+UGPIOC=23,0,0 ret:%d", ret);
+		goto error;
+	}
+
+	/* Let SIM settle */
+	k_sleep(MDM_CMD_TIMEOUT);
+#endif
+
+	/* UNC messages for registration */
+	ret = send_at_cmd(NULL, "AT+CREG=1", MDM_CMD_TIMEOUT);
+	if (ret < 0) {
+		LOG_ERR("AT+CREG=1 ret:%d", ret);
+		goto error;
+	}
+
+	/* HEX receive data mode */
+	ret = send_at_cmd(NULL, "AT+UDCONF=1,1", MDM_CMD_TIMEOUT);
+	if (ret < 0) {
+		LOG_ERR("AT+UDCONF=1 ret:%d", ret);
+	}
+
+	/* query modem info */
+	LOG_INF("Querying modem information");
+	ret = send_at_cmd(NULL, "ATI", MDM_CMD_TIMEOUT);
+	if (ret < 0) {
+		LOG_ERR("ATI ret:%d", ret);
+		goto error;
+	}
+
+	/* query modem IMEI */
+	ret = send_at_cmd(NULL, "AT+CGSN", MDM_CMD_TIMEOUT);
+	if (ret < 0) {
+		LOG_ERR("AT+CGSN ret:%d", ret);
+		goto error;
+	}
+
+	/* setup PDP context definition */
+
+	ret = send_at_cmd(NULL, "AT+CGDCONT=1,\"IP\",\""
+				CONFIG_MODEM_UBLOX_SARA_R4_APN "\"",
+			  MDM_CMD_TIMEOUT);
+	if (ret < 0) {
+		LOG_ERR("AT+CGDCONT ret:%d", ret);
+		goto error;
+	}
+
+	/* start functionality */
+	ret = send_at_cmd(NULL, "AT+CFUN=1", MDM_CMD_TIMEOUT);
+	if (ret < 0) {
+		LOG_ERR("AT+CFUN=1 ret:%d", ret);
+		goto error;
+	}
+
+	if (strlen(CONFIG_MODEM_UBLOX_SARA_R4_MANUAL_MCCMNO) > 0) {
+		/* use manual MCC/MNO entry */
+		ret = send_at_cmd(NULL, "AT+COPS=1,2,\""
+				  CONFIG_MODEM_UBLOX_SARA_R4_MANUAL_MCCMNO "\"",
+				  MDM_CMD_TIMEOUT);
+	} else {
+		/* register operator automatically */
+		ret = send_at_cmd(NULL, "AT+COPS=0,0", MDM_REGISTRATION_TIMEOUT);
+	}
+
+	if (ret < 0) {
+		LOG_ERR("AT+COPS ret:%d", ret);
+		goto error;
+	}
+
+	LOG_INF("Waiting for network");
+
+	/* wait for +CREG: 1 notification (20 seconds max) */
+	counter = 0;
+	while (counter++ < 20 && ictx.ev_creg != 1) {
+		k_sleep(K_SECONDS(1));
+	}
+
+	/* query modem RSSI */
+	modem_rssi_query_work(NULL);
+	k_sleep(MDM_WAIT_FOR_RSSI_DELAY);
+
+	counter = 0;
+	/* wait for RSSI < 0 and > -1000 */
+	while (counter++ < MDM_WAIT_FOR_RSSI_COUNT &&
+	       (ictx.mdm_ctx.data_rssi >= 0 ||
+		ictx.mdm_ctx.data_rssi <= -1000)) {
+		/* stop RSSI delay work */
+		k_delayed_work_cancel(&ictx.rssi_query_work);
+		modem_rssi_query_work(NULL);
+		k_sleep(MDM_WAIT_FOR_RSSI_DELAY);
+	}
+
+	if (ictx.mdm_ctx.data_rssi >= 0 || ictx.mdm_ctx.data_rssi <= -1000) {
+		retry_count++;
+		if (retry_count >= MDM_NETWORK_RETRY_COUNT) {
+			LOG_ERR("Failed network init.  Too many attempts!");
+			ret = -ENETUNREACH;
+			goto error;
+		}
+
+		LOG_ERR("Failed network init.  Restarting process.");
+		goto restart;
+	}
+
+	LOG_INF("Network is ready.");
+
+	/* Set iface up */
+	net_if_up(ictx.iface);
+
+error:
+	return;
+}
+
+static int modem_init(struct device *dev)
+{
+	int i, ret = 0;
+
+	ARG_UNUSED(dev);
+
+	/* check for valid pinconfig */
+	__ASSERT(sizeof(pinconfig) == MAX_MDM_CONTROL_PINS,
```

Comment:
Shouldn't `sizeof(pinconfig)` be `ARRAY_SIZE(pinconfig)` here? The size of pinconfig is larger than the number of pins because it's an array with a struct per pin.

Reply:
Huh.  I suppose I never turned on asserts and saw this fail.  But you are correct.